subscriptions/restHook.ts

```diff
@@ -0,0 +1,127 @@
+import axios from 'axios';
+import { SubscriptionEndpoint } from 'fhir-works-on-aws-routing/lib/router/validation/subscriptionValidator';
+import { groupBy } from 'lodash';
+import { makeLogger } from 'fhir-works-on-aws-interface';
+import { SubscriptionMatchMessage } from './types';
+import getAllowListedSubscriptionEndpoints from './allowList';
+import ensureAsyncInit from '../src/index';
+
+const logger = makeLogger();
+
+const SINGLE_TENANT_ALLOW_LIST_KEY = 'SINGLE_TENANT_ALLOW_LIST_KEY';
+
+export interface AllowListInfo {
+    allowList: (string | RegExp)[];
+    headerMap: { [key: string]: string[] };
+}
+
+const extractAllowListInfo = (subscriptionEndpoints: SubscriptionEndpoint[]): AllowListInfo => {
+    const allowList: (string | RegExp)[] = [];
+    const headerMap: { [key: string]: string[] } = {};
+    subscriptionEndpoints.forEach((allowEndpoint: SubscriptionEndpoint) => {
+        allowList.push(allowEndpoint.endpoint);
+        headerMap[allowEndpoint.endpoint.toString()] = allowEndpoint.headers || [];
+    });
+    return { allowList, headerMap };
+};
+
+async function getAllowListInfo(): Promise<{ [key: string]: AllowListInfo }> {
+    const originalAllowList = await getAllowListedSubscriptionEndpoints();
+    logger.debug(originalAllowList);
+    if (process.env.ENABLE_MULTI_TENANCY !== 'true') {
+        return { [SINGLE_TENANT_ALLOW_LIST_KEY]: extractAllowListInfo(originalAllowList) };
+    }
+    const allowListInfo: { [key: string]: AllowListInfo } = {};
+    const endpointsGroupByTenant: { [key: string]: SubscriptionEndpoint[] } = groupBy(
+        originalAllowList,
+        (allowEndpoint: SubscriptionEndpoint) => allowEndpoint.tenantId,
+    );
+    Object.entries(endpointsGroupByTenant).forEach(([key, value]) => {
+        allowListInfo[key] = extractAllowListInfo(value);
+    });
+    return allowListInfo;
+}
+
+/**
+ * Verify endpoint is allow listed
+ * Return allow list headers if endpoint is allow listed
+ * Throw error if endpoint is not allow listed
+ * @param allowListInfo
+ * @param endpoint
+ * @param tenantId
+ */
+const getAllowListHeaders = (
+    allowListInfo: { [key: string]: AllowListInfo },
+    endpoint: string,
+    tenantId?: string,
+): string[] => {
+    const getHeaders = ({ allowList, headerMap }: AllowListInfo): string[] => {
+        // eslint-disable-next-line no-restricted-syntax
+        for (const allowedEndpoint of allowList) {
+            if (allowedEndpoint instanceof RegExp && allowedEndpoint.test(endpoint)) {
+                return headerMap[allowedEndpoint.toString()];
+            }
+            if (allowedEndpoint === endpoint) {
+                return headerMap[allowedEndpoint];
+            }
+        }
+        throw new Error(`Endpoint ${endpoint} is not allow listed.`);
+    };
+
+    if (process.env.ENABLE_MULTI_TENANCY === 'true') {
```

Comment:
We should avoid accessing env variables deep within our code (I know FWoA does that a lot but it's a bad practice.). It makes it hard to know how the code behaves(e.g. looking at the index file it is hard to tell that `
exports.handler = sendRestHookNotification;` behaves differently with multi-tenancy)

I sugest to pass enableMultitenancy as a param instead. You may need to refactor into a class to "remember" the value of the env variable
```ts
const restHookHandler = new RestHookHandler({ enableMultitenancy: process.env.ENABLE_MULTI_TENANCY === 'true' });

export async function handleSQSevent(event: any) {
    return restHookHandler.sendRestHookNotification(event);
}
```

although passing the param on all function calls works as well
```ts
export async function handleSQSevent(event: any) {
    sendRestHookNotification(event, { enableMultitenancy: process.env.ENABLE_MULTI_TENANCY === 'true' });
}
```